src/master/develop/embedding.md

```diff
@@ -1,90 +1,172 @@
+## Embedding immudb SQL store
 
-```go	
-r, err = engine.QueryStmt("SELECT id, date, creditaccount, debitaccount, amount, description FROM journal WHERE amount > @value", map[string]interface{}{"value": 100}, true)
-```
+immudb provides you a immutable embedded SQL engine which keeps all history, is tamper-proof and can travel in time.
+The SQL engine is mounted on top of the embedded key value store.
 
-To iterate over a result set `r`, just fetch rows until there are no more entries. Every row has a `Values` member you can index to access the column:
+The following illustrative example showcase how to initialize the SQL engine, write and read data in the scope of a SQL transaction:
 
 ```go
-for {
-	row, err := r.Read()
-	if err == sql.ErrNoMoreRows {
-		break
-	}
+package main
+
+import (
+	"log"
+
+	"github.com/codenotary/immudb/embedded/sql"
+	"github.com/codenotary/immudb/embedded/store"
+)
+
+func handleErr(err error) {
 	if err != nil {
 		log.Fatal(err)
 	}
+}
+
+func main() {
+	// create/open immudb store at specified path
+	st, err := store.Open("data", store.DefaultOptions())
+	handleErr(err)
+	defer st.Close()
+
+	// initialize sql engine (specify a key-prefix to isolate generated kv entries)
+	engine, err := sql.NewEngine(st, sql.DefaultOptions().WithPrefix([]byte("sql")))
+	handleErr(err)
+
+	_, _, err = engine.Exec("CREATE DATABASE db1;", nil, nil)
+	handleErr(err)
+
+	// a sql tx is created and carried over next statements
+	sqltx, _, err := engine.Exec("BEGIN TRANSACTION;", nil, nil)
+	handleErr(err)
+
+	// set the database to use in the context of the ongoing sql tx
+	_, _, err = engine.Exec("USE DATABASE db1;", nil, sqltx)
+	handleErr(err)
+
+	// creates a table
+	_, _, err = engine.Exec(`
+		CREATE TABLE journal (
+			id INTEGER,
+			date TIMESTAMP,
+			creditaccount INTEGER,
+			debitaccount INTEGER,
+			amount INTEGER,
+			description VARCHAR,
+			PRIMARY KEY id
+		);`, nil, sqltx)
+	handleErr(err)
+
+	// insert some rows
+	_, _, err = engine.Exec(`
+		INSERT INTO journal (
+			id,
+			date,
+			creditaccount,
+			debitaccount,
+			amount,
+			description
+		) VALUES 
+			(1, NOW(), 100, 0, 4000, 'CREDIT'),
+			(2, NOW(), 0, 50, 4100, 'DEBIT')
+		;`, nil, sqltx)
+	handleErr(err)
+
+	// query data including ongoing and unconfirmed changes
+	rowReader, err := engine.Query(`
+			SELECT id, date, creditaccount, debitaccount, amount, description
+			FROM journal
+			WHERE amount > @value
+	`, map[string]interface{}{"value": 100}, sqltx)
+	handleErr(err)
+
+	// selected columns can be read from the rowReader
+	cols, err := rowReader.Columns()
+	handleErr(err)
+
+	for {
+		// iterate over result set
+		row, err := rowReader.Read()
+		if err == sql.ErrNoMoreRows {
+			break
+		}
+		handleErr(err)
+
+		// each row contains values for the selected columns
+		log.Printf("row: %v\n", row.Values[cols[0].Selector()].Value())
+	}
+
+	// close row reader
+	err = rowReader.Close()
```

Comment:
I'm wondering if we shouldn't do this in `defer` - this code will most likely be copied to other applications and could end up with a leak when the loop above errors out in `rowReader.Read`